src/main/java/htsjdk/samtools/SamInputResource.java

```diff
@@ -93,7 +93,7 @@ public static SamInputResource of(final File file) {
     /** Creates a {@link SamInputResource} reading from the provided resource, with no index. */
     public static SamInputResource of(final Path path) {
 
-        if (Files.isRegularFile(path) &&  Files.exists(path)) {
+        if (Files.isRegularFile(path)) {
```

Comment:
Was talking with @lbergelson just now, who pointed out that `Files.isRegularFile()` is also expensive over GCS (about as expensive as `Files.exists()`). We think it might be a good idea to check the scheme of the `Path` first before doing the `Files.isRegularFile()` check. If the Path has a scheme, and it's not `file://`, then create a `PathInputResource` unconditionally. If the Path has no scheme, or the scheme is `file://`, then do the `Files.isRegularFile()` check to decide whether to create a `PathInputResource` or a `FileInputResource`. Sound reasonable?

You can check the scheme of a `Path` via `path.toUri().getScheme()`. If you get an exception, then you can just catch it and fall back to the previous behavior.

Reply:
How about `path.getFileSystem() != FileSystems.getDefaultFileSystem()`?

Reply:
That seems like it would probably work.